src/GetDataCapableTrait.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+namespace Dhii\Data\Object;
+
+use Dhii\Util\String\StringableInterface as Stringable;
+use Exception as RootException;
+
+/**
+ * Functionality for data retrieval.
+ *
+ * @since [*next-version*]
+ */
+trait GetDataCapableTrait
+{
+    /**
+     * Retrieve data, all or by key.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable|null $key The key, for which to get the data,
+     *                                    or null to get all data as an array.
+     *
+     * @return mixed|array The value for the specified key, or all data as an key-value map.
+     */
+    protected function _getData($key = null)
+    {
+        if (!is_null($key) && !is_string($key) && !($key instanceof Stringable)) {
+            throw $this->_createInvalidArgumentException($this->__('Data key must be stringable'), null, null, $key);
+        }
+
+        $store = $this->_getDataStore();
+
+        // Return whole set
+        if (is_null($key)) {
+            return (array) $store;
+        }
+
+        $key = (string) $key;
+        if (!property_exists($store, $key)) {
+            throw $this->_createNotFoundException($this->__('Data key not found'), $key);
+        }
+
+        return $store->{$key};
+    }
+
+    /**
+     * Retrieves a pointer to the data store.
+     *
+     * @since [*next-version*]
+     *
+     * @return mixed The data store.
```

Comment:
Perhaps this should type-hint a return type of `object` since `_getData()` expect an object.